_posts/2021-03-09-bezier-fitting.md

```diff
@@ -0,0 +1,196 @@
+
+I believe there is no one perfect error metric, and the choice depends on the context. Fortunately, for smooth curves such as we're likely to find in fonts, the differences between these error metrics are subtle, and we can confidently choose the one that's easiest to reason about mathematically. For this blog post, that's the L2 norm.
+
+Two things more to say: for smooth curves + low error, leash is generally perpendicular. Also, normalized arc length (both engineer and Pierre move at a constant speed) is a good approximation, though strictly larger.
+
+## Stating the problem
+
+Our dog Pierre has a funny quirk: he is capable of moving only along a cubic bezier path. That said, he is otherwise fairly obedient, and always starts and ends at the same position as the engineer, as well as starting out and ending up in the same direction.
+
+Another way to state this is that our cubic bezier must match the source curve to G1. Then we want to find values for the parameters that minimizes the error norm. For a cubic bezier, the two parameters are the lengths of the two control arms; all other parameters are determined by the requirement to match the source curve.
+
+As a mathematical simplification, we can factor out translation, uniform scaling, and rotation, and just consider a curve that goes from (0, 0) to (1, 0), with given $\theta_0$ and $\theta_1$ angles.
+
+[image]
+
+Then, the control points of the bezier are $(0, 0)$, $(\delta_0 \cos \theta_0, \delta_0 \sin \theta_0)$, $(1 - \delta_1 \cos \theta_1, \delta_1 \sin \theta_1)$, $(1, 0)$.
+
+## Prior work
+
+Having stated the problem, we can take a look at the existing literature.
+
+* G2 geometric Hermite interpolation, de Boor. A bunch of new things here: a proof of O(n^6) scaling. The presence of up to three solutions. However, the constant factor on error is 5-10x, which is not great.
+
+* Two approaches from the Penner paper, center of mass and orthogonal distance fitting (ODF).
+
+* The Graphics Gems chapter.
+
+Also note, I addressed this problem in chapter 9 of my [thesis]. That approach gave good results but was very slow. The solution in this blog is strictly better.
+
+## Solving the problem
+
+We have two parameters to optimize. A good place to start is a heat map of the error over the entire two dimensional parameter space. Below is such a heatmap for the Euler spiral with deflections 0.31 and 0.35 radians at the endpoints. The horizontal axis is $\delta_0$, ie the length of the left control arm, and the vertical axis is $\delta_1$, the length of the right control arm.
```

Comment:
It might be useful to link out to the maths that explains how to get that euler curve segment? Most folks will never have even considered this curve type =D

Reply:
I linked Wikipedia, but it's not fantastic. We need an "Euler spiral segment primer" like the bezier primer :)